### PR TITLE
Bug 2002368: acccount for image api returning invalid on imagestream create based on allowed/blocked registry settings

### PR DESCRIPTION
**Pull request by gabemontero**
Fixes https://github.com/openshift/cluster-samples-operator/issues/385

### CODE REVIEW

**Comment by gabemontero** [comment]
/hold

I can get a bit more sophisticated / robust with this.

**Comment by gabemontero** [comment]
aws flake

/test e2e-aws-upgrade

**Review by dperaza4dustbit — APPROVED** [review]
/lgtm

**Comment by gabemontero** [comment]
yeah I'm going to pivot from our prior conversations @dperaza4dustbit 

I'm going to watch allowed/blocked registries, and if we are blocked, I'm going to bootstrap as removed, similar to what we do with ipv6/proxy/disconnected .... a "separate cluster config" situation makes samples install untenable.

I'll add a new reason code when we are removed, and I believe the existing alert we have there should be sufficient.

With that, I don't think there would be any future work for you team to do wrt new metrics and alerts.

This will require a doc update with @rolfedh (which I just mentioned in scrum with him).  We'll need to work with him here in this PR, perhaps with a Jira on his board, to track.

/assign @rolfedh

File: pkg/stub/imagestreams.go
Comment (stbenjam) [inline pkg/stub/imagestreams.go:179]: This is a pre-existing problem but this line causes panics, imagestreamInCluster is guaranteed to be nil (line 170 checks)

https://bugzilla.redhat.com/show_bug.cgi?id=2003683

Comment (gabemontero) [reply pkg/stub/imagestreams.go]: thanks for heads up @stbenjam  ... I've added a commit in this PR for now, but have also pinged @dperaza4dustbit about this 

we'll either address here with that commit , or he can do a separate PR, based on his taking ownership of samples (we are in the middle of the transition this week).

**Comment by gabemontero** [comment]
/hold

the new global image config check needs some work

File: manifests/010-prometheus-rules.yaml
Comment (gabemontero) [inline manifests/010-prometheus-rules.yaml:98]: potential doc updates for these changes @dperaza4dustbit

File: pkg/stub/config.go
Comment (gabemontero) [inline pkg/stub/config.go:303]: A doc update around this new reason for our config invalid condition is probably warranted @dperaza4dustbit

File: pkg/stub/handler.go
Comment (gabemontero) [inline pkg/stub/handler.go:422]: doc update around this @dperaza4dustbit

**Comment by gabemontero** [comment]
just the okd upgrade ... non related failure

/skip

**Comment by coreydaley** [comment]
@gabemontero I've got this in my queue to look at in the morning.

File: README.md
Comment (coreydaley) [inline README.md:137]:
```suggestion
are not managed by the samples operator.  These are special imagestreams that point to images in the install payload, and their creation and updates are handled by the 
```

File: README.md
Comment (coreydaley) [inline README.md:160]:
```suggestion
- Delete the samples resource, `oc delete config.samples cluster`, config map/secrets in operator namespace, imagestreams and templates in openshift namespace
```

File: pkg/stub/handler.go
Comment (coreydaley) [inline pkg/stub/handler.go:312]:
```suggestion
	// the multi tag, highly concurrent nature of creating / updating;
```

File: pkg/stub/handler.go
Comment (coreydaley) [inline pkg/stub/handler.go:314]:
```suggestion
	// returns if a registry is missing from the images allowed registry list, or is explicitly
```

File: pkg/stub/handler.go
Comment (coreydaley) [inline pkg/stub/handler.go:361]:
```suggestion
						logrus.Printf(("the allowed registries domain %s allows imagestream creation access for search param %s", rl.DomainName, name)
```

Comment (coreydaley) [reply pkg/stub/handler.go:361]: This could be updated throughout the pull request. There shouldn't be a reason to create the `msg` variable and then pass it into `Printf` or `Errorf` unless it is going to be used again somewhere else.

File: pkg/stub/handler.go
Comment (coreydaley) [inline pkg/stub/handler.go:438]:
```suggestion
		return false, errors.New("proxy is not configured")
```

File: pkg/stub/handler.go
Comment (coreydaley) [inline pkg/stub/handler.go:352]: should `registry.redhat.io` be hard coded here? Or should it use the name of the registry that is passed in as `name`?

File: pkg/stub/handler.go
Comment (coreydaley) [inline pkg/stub/handler.go:354]: I don't think you need an else block here

File: pkg/stub/handler.go
Comment (coreydaley) [inline pkg/stub/handler.go:383]: Do we need a section to check for a partial match of domain name for blocked registries also, like we do above for allowed registries?

File: pkg/stub/handler.go
Comment (coreydaley) [inline pkg/stub/handler.go:353]: Would it make sense to just have this section in the `wait.PollImmediate` instead of the whole logic for this function?

File: pkg/stub/handler.go
Comment (coreydaley) [inline pkg/stub/handler.go:340]: Is it documented anywhere what order these rules are processed in? Does being specifically allowed override being specifically denied? Does a full match of an allowed registry override a partial match of a blocked registry, etc?

**Review by coreydaley — CHANGES_REQUESTED** [review]

File: pkg/stub/handler.go
Comment (gabemontero) [reply pkg/stub/handler.go:361]: yeah there were some cases where I was both creating a msg for a log as well as for an error to return

in those cases I'll try to address

in the cases where both are in fact occurring, I'll keep the pattern of defining the string once

File: pkg/stub/handler.go
Comment (gabemontero) [reply pkg/stub/handler.go:438]: where is `errors` suppose to come from ? ... my IDE package searches are not turning anything up

File: pkg/stub/handler.go
Comment (gabemontero) [reply pkg/stub/handler.go:352]: yep

File: pkg/stub/handler.go
Comment (gabemontero) [reply pkg/stub/handler.go:354]: yep

File: pkg/stub/handler.go
Comment (coreydaley) [reply pkg/stub/handler.go:438]: https://pkg.go.dev/errors

Comment (coreydaley) [reply pkg/stub/handler.go:438]: Part of the Go standard library

File: pkg/stub/handler.go
Comment (coreydaley) [reply pkg/stub/handler.go:352]: > yep

Yep to which one? :)

File: pkg/stub/handler.go
Comment (gabemontero) [reply pkg/stub/handler.go:383]: nope the domain thing only exists for allowed

then it is explicit registry listing for the stuff under RegistrySource

File: pkg/stub/handler.go
Comment (gabemontero) [reply pkg/stub/handler.go:353]: sorry not following / can't parse

Comment (coreydaley) [reply pkg/stub/handler.go:353]: Would it make sense for only the Get() to be inside of the wait.PollImmediate, instead of all the code within the function being in there? Probably doesn't really matter honestly, just a thought.

File: pkg/stub/handler.go
Comment (gabemontero) [reply pkg/stub/handler.go:340]: order does not matter .... a problem with any of them, we punt

even if they did something like putting something in both allowed and blocked ... too bad, we punt until you fix your config so it makes sense

in any event, nuances like that should be handled by the owners of the config api itself ... they should have validation to catch that 

ultimately the doc updates I'll work out with @rolfedh will get into that

**Comment by gabemontero** [comment]
updated I did make are in separate commits for now @coreydaley .... I'll squash after you sign off on those, plus we sort out remaining questions

also, ideally we do not merge until I hear again from @dperaza4dustbit on these more recent changes ... he lgtmed an earlier version

File: pkg/stub/handler.go
Comment (gabemontero) [reply pkg/stub/handler.go]: switch to param name

**Comment by gabemontero** [comment]
don't know why I couldn't find "errors" before ... switch will be up momentarily @coreydaley

File: pkg/stub/handler.go
Comment (gabemontero) [reply pkg/stub/handler.go]: ah ... you are right that is better .... update up in a couple

**Comment by gabemontero** [comment]
round 2 updates up @coreydaley

**Comment by gabemontero** [comment]
known unit flake with okd version .. all other okd's passed

/skip

/retest

File: pkg/stub/config.go
Comment (dperaza4dustbit) [reply pkg/stub/config.go:303]: got it

File: pkg/stub/handler.go
Comment (dperaza4dustbit) [reply pkg/stub/handler.go:430]: got it

**Comment by gabemontero** [comment]
great thanks @coreydaley @dperaza4dustbit 

I've squashed the review commits from @coreydaley 's reviews

if I could get one more lgtm from one of you, we'll go from there

about to ping @rolfedh on slack now to get the doc updates going

**Comment by gabemontero** [comment]
unrelated flakes in okd build test failure; sig-builds passed

/skip

**Comment by gabemontero** [comment]
unrelated sig-instrumentation flake image eco

/test e2e-aws-image-ecosystem

**Comment by gabemontero** [comment]
when it is time to cherrypick, the 4.9 and other z streams will need both this PR and https://github.com/openshift/cluster-samples-operator/pull/397

**Comment by gabemontero** [comment]
/assign @jitendar-singh 

ideally you apply the qe approved label to this PR

/assign @rolfedh 

ideally you apply the docs approved label to this PR, where the input I gave you at https://docs.google.com/document/d/1fN_-2_ZPC6DQBIYJkrlqL_0i9GSVbqJ8YyhB8hRns-o/edit is the starting point you need

**Pull request merged**
